src/clients/patreon/patreonAgent.ts

```diff
@@ -0,0 +1,235 @@
+
+    const existingPatreon = await this.getExistingPatreon(usernameLower);
+
+    if (!existingPatreon) {
+      return null;
+    }
+
+    const isActivePatreon = !this.hasExpired(
+      existingPatreon.currentPledgeExpiryDate,
+    );
+
+    return {
+      patreonUserId: existingPatreon.patreonUserId,
+      patreonUsersName: existingPatreon.patreonUsersName,
+      isActivePatreon,
+      amountCents: existingPatreon.amountCents,
+    };
+  }
+
+  // This path is hit whenever user clicks link to Patreon button
+  public async linkUserToPatreon(
+    usernameLower: string,
+    code: string,
+  ): Promise<PatreonDetails> {
+    // Grab user tokens
+    const tokens = await this.patreonController.getTokens(code);
+
+    // Grab member details from Patreon with token
+    const patronDetails = await this.patreonController.getPatronDetails(
+      tokens.userAccessToken,
+    );
+
+    // Grab Patreon document from MongoDB
+    const existingPatreon = await this.getExistingPatreon(usernameLower);
+
+    // Do not let more than one patreon be used for same user
+    if (
+      existingPatreon &&
+      existingPatreon.patreonUserId !== patronDetails.patreonUserId
+    ) {
+      throw new Error(
+        'Attempted to upload a second Patreon for the same user.',
+      );
+    }
+
+    // Do not let one patreon be used for more than one user
+    const patreonAccountInUse = await patreonId.findOne({
+      patreonUserId: patronDetails.patreonUserId,
+    });
+    if (
+      patreonAccountInUse &&
+      patreonAccountInUse.proavalonUsernameLower !== usernameLower
+    ) {
+      throw new Error(
+        'Attempted to upload a used Patreon for more than one user.',
+      );
+    }
+
+    let result: PatreonDetails;
+
+    if (patronDetails.patreonMemberDetails) {
+      // They are a current member
+      result = await this.updateCurrentPatreonMember(
+        existingPatreon,
+        patronDetails,
+        usernameLower,
+        tokens,
+      );
+    } else {
+      // They are not a current member to the Patreon page
+      result = await this.updateCurrentNonPatreonMember(
+        existingPatreon,
+        tokens,
+        usernameLower,
+        patronDetails.patreonUserId,
+      );
+    }
+
+    console.log(
+      `Successfully linked Patreon account: proavalonUsernameLower="${usernameLower}" patreonUserId="${patronDetails.patreonUserId}" patreonName="${patronDetails.patreonUsersName}" isActivePatreon="${result.isActivePatreon}" amountCents="${result.amountCents}"`,
+    );
+
+    return result;
+  }
+
+  private async updateCurrentPatreonMember(
+    existingPatreon: any,
+    patronDetails: any,
+    usernameLower: string,
+    tokens: PatreonUserTokens,
+  ): Promise<PatreonDetails> {
+    const amountCents =
+      patronDetails.patreonMemberDetails.currently_entitled_amount_cents;
+    const lastChargeDate = new Date(
+      patronDetails.patreonMemberDetails.last_charge_date,
```

Comment:
```suggestion
      patronDetails.patreonMemberDetails.last_charge_date, // ISO date string
```

Reply:
On second thought, let's update getPatronDetails in the controller to return a direct type with an interface that does validation on this kind of last_charge_date. That way the moment it comes back to the agent, we know for sure these values exist.